client/app/pages/users/UsersList.jsx

```diff
@@ -0,0 +1,251 @@
+          {user.is_invitation_pending && <span className="label label-tag-archived">Invitation Pending</span>}
+        </div>
+      </div>
+    ), {
+      title: 'Name',
+      field: 'name',
+      width: null,
+      className: 'p-l-0',
+    }),
+    Columns.custom.sortable((text, user) => map(user.groups, group => (
+      <a key={'group' + group.id} className="label label-tag" href={'groups/' + group.id}>{group.name}</a>
+    )), {
+      title: 'Groups',
+      field: 'groups',
+    }),
+    Columns.timeAgo.sortable({ title: 'Joined', field: 'created_at' }),
+    Columns.timeAgo.sortable({ title: 'Last Active At', field: 'active_at' }),
+    Columns.custom((text, user, context) => {
+      if (user.id !== currentUser.id) {
+        if (user.is_invitation_pending) {
+          return (
+            <button type="button" className="btn btn-default btn-block" onClick={event => context.deleteUser(event, user)}>Delete</button>
+          );
+        }
+        return user.is_disabled ? (
+          <button type="button" className="btn btn-primary btn-block" onClick={event => context.enableUser(event, user)}>Enable</button>
+        ) : (
+          <button type="button" className="btn btn-default btn-block" onClick={event => context.disableUser(event, user)}>Disable</button>
+        );
+      }
+      return null;
+    }, {
+      className: 'p-r-0',
+      isAvailable: () => currentUser.isAdmin,
+    }),
+  ];
+
+  constructor(props) {
+    super(props);
+
+    const wrapUserAction = action => (
+      (event, user) => {
+        event.preventDefault();
+        event.stopPropagation();
+        // User service will handle errors, no need to do it here
+        action(user).then(() => {
+          this.controller.update();
+        });
+      }
+    );
+
+    this.actions = {
+      enableUser: wrapUserAction(User.enableUser),
+      disableUser: wrapUserAction(User.disableUser),
+      deleteUser: wrapUserAction(User.deleteUser),
+    };
+
+    const resource = User.query.bind(User);
+
+    this.controller = new LiveItemsList({
+      getRequest: (request) => {
+        if (this.props.currentPage === 'disabled') {
+          request.disabled = true;
+        }
+        return request;
+      },
+      doRequest: LiveItemsList.createFetcher(resource, User),
+      onChange: ({ state }) => this.setState(state),
+    });
+    this.state = this.controller.state;
+
+    this.onTableRowClick = (event, item) => navigateTo('users/' + item.id);
+  }
+
+  componentDidMount() {
+    this.controller.update();
+  }
+
+  renderPageHeader(isAdminView) {
+    return isAdminView ? (
+      // Admin
+      <div className="m-b-10">
+        <a
+          href="users/new"
+          className={classNames('btn', 'btn-default', 'm-b-10', { disabled: !policy.isCreateUserEnabled() })}
+        >
+          <i className="fa fa-plus m-r-5" />
+          New User
+        </a>
+        <DynamicComponent is="UsersListExtra" />
```

Comment:
I tried really hard, but I have no idea what this does..

Reply:
Leave this for @arikfr

Reply:
It's a replacement for [this](https://github.com/getredash/redash/blob/master/client/app/pages/users/list.html#L7).

Reply:
I don't understand it but I'll leave it be 👍🏻